tests/unit/optimization/SGDMinimizer_unittest.h

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (c) The Shogun Machine Learning Toolbox
+ * Written (W) 2015 Wu Lin
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * The views and conclusions contained in the software and documentation are those
+ * of the authors and should not be interpreted as representing official policies,
+ * either expressed or implied, of the Shogun Development Team.
+ *
+ */
+#include <shogun/lib/config.h>
+#ifdef HAVE_EIGEN3
+#ifndef SGDMINIMIZER_UNITTEST_H
+#define SGDMINIMIZER_UNITTEST_H
+#include <shogun/optimization/FirstOrderStochasticCostFunction.h>
+#include <shogun/lib/SGMatrix.h>
+#include <shogun/lib/SGVector.h>
+using namespace shogun;
+
+class CRegressionExample;
+
+class CMyStochasticCostFunction: public CFirstOrderStochasticCostFunction
+{
+public:
+	CMyStochasticCostFunction();
+	virtual ~CMyStochasticCostFunction();
+	void set_target(CRegressionExample *obj);
+	virtual float64_t get_cost();
+	virtual CMap<TParameter*, SGVector<float64_t> >* obtain_variable_reference();
+	virtual CMap<TParameter*, SGVector<float64_t> >* get_gradient();
+	virtual void begin_sample();
+	virtual bool next_sample();
+	virtual const char* get_name() const {return "MyStochasticCostFunction";}
+private:
+	index_t m_idx;
+	void init();
+	CRegressionExample* m_obj;
+};
+
+
+class CRegressionExample: public CSGObject
```

Comment:
this is a problem we try to solve.
since we do not do multiply inheritance, we need to define its cost function separately.